patroni/dcs/__init__.py

```diff
@@ -172,28 +193,46 @@ def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
 
 class Member(NamedTuple):
     """Immutable object (namedtuple) which represents single member of PostgreSQL cluster.
+
+    .. note::
+        There are two mandatory keys in a data:
+
+        ``conn_url``: connection string containing host, user and password which could be used to access this member.
+        ``api_url``: REST API url of patroni instance
```

Comment:
"Noteworthy"?

Reply:
When Patroni updates the member key it always writes conn_url and api_url, but what if the key was updated by someone else? We want to be resilient for such unexpected scenarios.

Reply:
I don't understand this. So `Member` is read from DCS during `dcs.touch_member`, and if it's different to `data` passed into it, then the data constructed in `ha.touch_member` is written (looking at the etcd3 implementation). The values for each of these 2 parameters are taken from `state_handler.connection_string` and `api.connection_string` in the `ha.touch_member`. I don't see where the existing data held in DCS is kept/merged.

Do you mean someone who wrote to the DCS `/member/member` key outside of Patroni? Who would do that? And why would that be considered something to honour?

I am thinking that for this comment it's worth mentioning that those keys are assembled from active connection information when data is written to DCS.

Also, why doesn't touch_member create a `Member` instance instead of constructing a `Dict`?

Thanks for any pointers!

Reply:
> So Member is read from DCS during dcs.touch_member

No, touch_member() write member key. Reading happens somewhere inside in get_cluster().

> and if it's different to data passed into it, then the data constructed in ha.touch_member is written (looking at the etcd3 implementation)

This is unrelated and just implementation detail of very specific DCS, because the member key is connected with the session and will not expire while the session exists. Therefore we can skip the update if value didn't changed.
But, with for example etcd v2 the member key is updated every time.

> Do you mean someone who wrote to the DCS /member/member key outside of Patroni? Who would do that? And why would that be considered something to honour?

Exactly, something different from Patroni could change arbitrary keys in DCS and we better be resilient to it.

> Also, why doesn't touch_member create a Member instance instead of constructing a Dict?

Because touch_member() only updates one key that represents current node in DCS. When we read the cluster state we are reading **all** member keys, which happens on all nodes.

Reply:
> How are we resilient to it? 

Until Patroni on respective node overwrites it.
That is, we don't want to suddenly crash because can't deserialize something.
But, we still create the Member object because the node that performs update of the member key sometimes should know that the key exists and should be removed before creating a new one.
For example in Zookeeper and Consul keys could be bound to a session and it is not possible (or there is no reliable way of doing) to change the session for the key.

Reply:
Ah right, got it. So you're referring to the read of DCS being resilient to changes UNTIL we overwrite it with sane values.

Reply:
Anything is possible. Maybe the broken member key will never be updated, we just need to make sure that it will not affect normal cluster operations. For example we don't want to create a replication slot for it.

Reply:
How about this:

```suggestion
    .. note::
        These two keys in data are always written to the DCS, but care is taken to maintain consistency and resilience
        from data that is read:
        
        ``conn_url``: connection string containing host, user and password which could be used to access this member.
        ``api_url``: REST API url of patroni instance
```